zstd/seqdec_amd64.s

```diff
@@ -0,0 +1,604 @@
+
+	MOVQ    mlState, DX // DX = newState()
+	SHRQ    $16, DX
+	MOVWQZX DX, DX
+
+	ADDQ BX, DX            // DX += lowBits()
+	ANDQ $maxTableMask, DX // DX &= maxTableMask
+
+	// mlState = llTable[...]
+	MOVQ decodeAsmContext_mlTable(R15), AX
+	MOVQ (AX)(DX*8), mlState
+
+	// lowBits := getBits(mlState.nbBits())
+	MOVBQZX ofState, AX // AX = nbBits() -- note: nbBits is the lowest byte of state
+
+	// BX = lowBits
+
+#ifdef GOAMD64_v3
+	LEAQ  (br_bits_read)(AX*1), CX
+	MOVQ  br_value, BX
+	MOVQ  CX, br_bits_read
+	ROLQ  CL, BX
+	BZHIQ AX, BX, BX
+
+#else
+	MOVQ    br_bits_read, CX // BX = ((br_value << br_bits_read) >> (64 - AX))
+	ADDQ    AX, br_bits_read
+	MOVQ    br_value, BX
+	SHLQ    CL, BX
+	MOVQ    AX, CX
+	XORQ    AX, AX
+	NEGQ    CX
+	SHRQ    CL, BX
+	TESTQ   CX, CX
+	CMOVQEQ AX, BX
+	MOVWQZX BX, BX
+
+#endif
+
+	MOVQ    ofState, DX // DX = newState()
+	SHRQ    $16, DX
+	MOVWQZX DX, DX
+
+	ADDQ BX, DX            // DX += lowBits()
+	ANDQ $maxTableMask, DX // DX &= maxTableMask
+
+	// mlState = llTable[...]
+	MOVQ decodeAsmContext_ofTable(R15), AX
+	MOVQ (AX)(DX*8), ofState
+
+skip_update:
+
+	// mo = s.adjustOffset(mo, ll, moB)
+	// prepare args for sequenceDecs_adjustOffsets
+	MOVQ    s+0(FP), DI
+	MOVQ    seqVals_mo(seqs), BX            // mo
+	MOVQ    seqVals_ll(seqs), CX            // ll
+	MOVQ    R14, DX                         // moB (from the ofState before its update)
+	SHRQ    $8, DX
+	MOVBQZX DL, DX
+	LEAQ    sequenceDecs_prevOffset(DI), DI // DI = &s.prevOffset[0]
+
+	// if offsetB > 1 {
+	//     s.prevOffset[2] = s.prevOffset[1]
+	//     s.prevOffset[1] = s.prevOffset[0]
+	//     s.prevOffset[0] = offset
+	//     return offset
+	// }
+	CMPQ DX, $1
+	JBE  offsetB_1_or_0
+
+	MOVQ 8(DI), AX
+	MOVQ AX, 16(DI) // s.prevOffset[2] = s.prevOffset[1]
+	MOVQ 0(DI), AX
+	MOVQ AX, 8(DI)  // s.prevOffset[1] = s.prevOffset[0]
+	MOVQ BX, 0(DI)  // s.prevOffset[0] = offset
+
+	MOVQ BX, BX
+	JMP  check_triple
+
+offsetB_1_or_0:
+	// if litLen == 0 {
+	//     offset++
+	// }
+	LEAQ    1(BX), AX // AX = offset + 1
+	TESTQ   CX, CX
+	CMOVQEQ AX, BX    // offset++ if litLen == 0
+
+	// if offset == 0 {
+	//     return s.prevOffset[0]
+	// }
+	TESTQ BX, BX
+	JNZ   offset_nonzero
+	MOVQ  0(DI), BX
+	JMP   check_triple
+
+offset_nonzero:
+	// Note: at this point CX (litLen) and DX (offsetB) are free
+
+	// var temp int
+	// if offset == 3 {
+	//     temp = s.prevOffset[0] - 1
+	// } else {
+	//     temp = s.prevOffset[offset]
+	// }
+	//
+	// this if got transformed into:
+	//
+	// ofs   := offset
+	// shift := 0
+	// if offset == 3 {
+	//     ofs   = 0
+	//     shift = -1
+	// }
+	// temp := s.prevOffset[ofs] + shift
+
+	MOVQ    BX, CX   // ofs_other   = offset
+	XORQ    DX, DX   // shift_other = 0
+	XORQ    R14, R14 // ofs_3       = 0
+	MOVQ    $-1, R15 // shift_3     = -1
+	CMPQ    BX, $3
+	CMOVQEQ R14, CX
+	CMOVQEQ R15, DX
+
+	ADDQ 0(DI)(CX*8), DX // DX is temp
+
+	// if temp == 0 {
+	//     temp = 1
+	// }
+	MOVQ    $1, CX
+	CMOVQEQ CX, DX
```

Comment:
CMOV here is likely slower since it can be 100% predicted (this is likely an error)

Reply:
So, do you suggest it should be a jump?

Reply:
Yeah, TEST and JNZ.

Reply:
CMOV tends to only be better when the branch is close to 50/50. This will be 100/0, since only broken input will trigger it.

Reply:
OK, changed.

Reply:
Still here AFAICT

Reply:
Sorry, I forgot to push the changes.